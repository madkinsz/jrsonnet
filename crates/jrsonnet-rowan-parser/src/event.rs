use std::mem;

use rowan::{GreenNode, GreenNodeBuilder, Language};

use crate::{
	lex::{Lang, Lexeme, SyntaxKind},
	parser::{Parse, SyntaxError},
};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Event {
	Start {
		kind: SyntaxKind,
		forward_parent: Option<usize>,
	},
	Token,
	Finish,
	Placeholder,
	Error(SyntaxError),
}

pub(super) struct Sink<'i> {
	pub builder: GreenNodeBuilder<'static>,
	lexemes: &'i [Lexeme<'i>],
	offset: usize,
	events: Vec<Event>,
	pub errors: Vec<SyntaxError>,
}

impl<'i> Sink<'i> {
	pub(super) fn new(events: Vec<Event>, lexemes: &'i [Lexeme<'i>]) -> Self {
		Self {
			builder: GreenNodeBuilder::new(),
			lexemes,
			offset: 0,
			events,
			errors: vec![],
		}
	}

	pub(super) fn finish(mut self) -> Parse {
		for idx in 0..self.events.len() {
			match mem::replace(&mut self.events[idx], Event::Placeholder) {
				Event::Start {
					kind,
					forward_parent,
				} => {
					let mut kinds = vec![kind];

					let mut idx = idx;
					let mut forward_parent = forward_parent;

					// Walk through the forward parent of the forward parent, and the forward parent
					// of that, and of that, etc. until we reach a StartNode event without a forward
					// parent.
					while let Some(fp) = forward_parent {
						idx += fp;

						forward_parent = if let Event::Start {
							kind,
							forward_parent,
						} = mem::replace(&mut self.events[idx], Event::Placeholder)
						{
							kinds.push(kind);
							forward_parent
						} else {
							unreachable!()
						};
					}

					for kind in kinds.into_iter().rev() {
						self.builder.start_node(Lang::kind_to_raw(kind));
					}
				}
				Event::Token => self.token(),
				Event::Finish => {
					self.builder.finish_node();
				}
				Event::Placeholder => {}
				Event::Error(e) => {
					self.errors.push(e);
				}
			}
			self.skip_whitespace();
		}

		Parse {
			green_node: self.builder.finish(),
			errors: self.errors,
		}
	}
	fn token(&mut self) {
		let lexeme = self.lexemes[self.offset];
		self.builder
			.token(Lang::kind_to_raw(lexeme.kind), lexeme.text);
		self.offset += 1;
	}
	fn skip_whitespace(&mut self) {
		while let Some(lexeme) = self.lexemes.get(self.offset) {
			if !lexeme.kind.is_trivia() {
				break;
			}

			self.token();
		}
	}
}
