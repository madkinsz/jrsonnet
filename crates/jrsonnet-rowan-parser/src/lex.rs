use crate::string_block::lex_str_block_test;
use core::ops::Range;
use logos::Logos;
use rowan::{Checkpoint, TextRange, TextSize};
use std::{convert::TryFrom, iter::Peekable};

#[derive(Logos, Debug, PartialEq, Hash, Eq, PartialOrd, Ord, Clone, Copy)]
#[repr(u16)]
pub enum SyntaxKind {
	#[token("assert")]
	KeywordAssert = 0,

	#[token("else")]
	KeywordElse,

	#[token("error")]
	KeywordError,

	#[token("false")]
	KeywordFalse,

	#[token("for")]
	KeywordFor,

	#[token("function")]
	KeywordFunction,

	#[token("if")]
	KeywordIf,

	#[token("import")]
	KeywordImport,

	#[token("importstr")]
	KeywordImportStr,

	#[token("local")]
	KeywordLocal,

	#[token("null")]
	KeywordNull,

	#[token("tailstrict")]
	KeywordTailStrict,

	#[token("then")]
	KeywordThen,

	#[token("self")]
	KeywordSelf,

	#[token("super")]
	KeywordSuper,

	#[token("true")]
	KeywordTrue,

	#[regex(r"[_a-zA-Z][_a-zA-Z0-9]*")]
	Ident,

	#[regex(r"(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?")]
	Number,

	#[regex(r"(?:0|[1-9][0-9]*)\.[^0-9]")]
	ErrorNumJunkAfterDecimalPoint,

	#[regex(r"(?:0|[1-9][0-9]*)(?:\.[0-9]+)?[eE][^+\-0-9]")]
	ErrorNumJunkAfterExponent,

	#[regex(r"(?:0|[1-9][0-9]*)(?:\.[0-9]+)?[eE][+-][^0-9]")]
	ErrorNumJunkAfterExponentSign,

	#[token("{")]
	SymbolLeftBrace,

	#[token("}")]
	SymbolRightBrace,

	#[token("[")]
	SymbolLeftBracket,

	#[token("]")]
	SymbolRightBracket,

	#[token(",")]
	SymbolComma,

	#[token(".")]
	SymbolDot,

	#[token("(")]
	LParen,

	#[token(")")]
	RParen,

	#[token(";")]
	SymbolSemi,
	#[token(":")]
	SymbolColon,

	#[token("$")]
	SymbolDollar,

	#[token("*")]
	OpMul,
	#[token("/")]
	OpDiv,
	#[token("%")]
	OpMod,
	#[token("+")]
	OpPlus,
	#[token("-")]
	OpMinus,
	#[token("<<")]
	OpShiftLeft,
	#[token(">>")]
	OpShiftRight,
	#[token("<")]
	OpLessThan,
	#[token(">")]
	OpGreaterThan,
	#[token("<=")]
	OpLessThanOrEqual,
	#[token(">=")]
	OpGreaterThanOrEqual,
	#[token("==")]
	OpEqual,
	#[token("!=")]
	OpNotEqual,
	#[token("&")]
	OpBitAnd,
	#[token("^")]
	OpBitXor,
	#[token("|")]
	OpBitOr,
	#[token("&&")]
	OpAnd,
	#[token("||")]
	OpOr,
	#[token("in")]
	OpIn,
	#[token("!")]
	OpNot,
	#[token("~")]
	OpBitNegate,
	#[token("=")]
	SymbolAssign,

	#[regex("\"(?s:[^\"\\\\]|\\\\.)*\"")]
	StringDoubleQuoted,

	#[regex("'(?s:[^'\\\\]|\\\\.)*'")]
	StringSingleQuoted,

	#[regex("@\"(?:[^\"]|\"\")*\"")]
	StringDoubleVerbatim,

	#[regex("@'(?:[^']|'')*'")]
	StringSingleVerbatim,

	#[regex(r"\|\|\|", lex_str_block_test)]
	StringBlock, //(StringBlockToken),

	#[regex("\"(?s:[^\"\\\\]|\\\\.)*")]
	ErrorStringDoubleQuotedUnterminated,

	#[regex("'(?s:[^'\\\\]|\\\\.)*")]
	ErrorStringSingleQuotedUnterminated,

	#[regex("@\"(?:[^\"]|\"\")*")]
	ErrorStringDoubleVerbatimUnterminated,

	#[regex("@'(?:[^']|'')*")]
	ErrorStringSingleVerbatimUnterminated,

	#[regex("@[^\"'\\s]\\S+")]
	ErrorStringMissingQuotes,

	#[token("/*/")]
	ErrorCommentTooShort,

	#[regex(r"/\*([^*]|\*[^/])+")]
	ErrorCommentUnterminated,

	#[regex(r"[ \t\n\r]+")]
	Whitespace,

	#[regex(r"//[^\r\n]*(\r\n|\n)?")]
	SingelLineSlashComment,

	#[regex(r"#[^\r\n]*(\r\n|\n)?")]
	SingleLineHashComment,

	#[regex(r"/\*([^*]|\*[^/])*\*/")]
	MultiLineComment,

	#[error]
	Error,

	ErrorPositionalAfterNamed,

	Literal,
	Expr,
	Array,
	ArrayElem,
	Object,
	Field,

	CompspecFor,
	CompspecIf,

	Slice,
	FieldAccess,
	ObjectApply,
	FunctionCall,
	FunctionDef,
	BodyDef,

	BinOp,
	UnaryOp,
	Local,
	ExprError,
	ExprAssert,
	ExprImport,

	DefParam,
	DefParams,

	DefArgs,
	DefNamedArg,
	DefPositionalArg,

	Parened,

	Root,
}

impl SyntaxKind {
	pub fn is_trivia(self) -> bool {
		matches!(
			self,
			Self::Whitespace
				| Self::MultiLineComment
				| Self::SingelLineSlashComment
				| Self::SingleLineHashComment
		)
	}
}

pub struct Lexer<'a> {
	inner: logos::Lexer<'a, SyntaxKind>,
}

impl<'a> Lexer<'a> {
	pub fn new(input: &'a str) -> Self {
		Self {
			inner: SyntaxKind::lexer(input),
		}
	}
}

impl<'a> Iterator for Lexer<'a> {
	type Item = Lexeme<'a>;

	fn next(&mut self) -> Option<Self::Item> {
		let kind = self.inner.next()?;
		let text = self.inner.slice();

		Some(Self::Item {
			kind,
			text,
			range: {
				let Range { start, end } = self.inner.span();

				TextRange::new(
					TextSize::try_from(start).unwrap(),
					TextSize::try_from(end).unwrap(),
				)
			},
		})
	}
}

#[derive(Clone, Copy)]
pub struct Lexeme<'i> {
	pub kind: SyntaxKind,
	pub text: &'i str,
	pub range: TextRange,
}

pub fn lex(input: &str) -> Vec<Lexeme<'_>> {
	Lexer::new(input).collect()
}

impl From<SyntaxKind> for rowan::SyntaxKind {
	fn from(kind: SyntaxKind) -> Self {
		Self(kind as u16)
	}
}

use SyntaxKind::*;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Lang {}
impl rowan::Language for Lang {
	type Kind = SyntaxKind;
	fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
		assert!(raw.0 <= Root as u16);
		unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
	}
	fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
		kind.into()
	}
}
