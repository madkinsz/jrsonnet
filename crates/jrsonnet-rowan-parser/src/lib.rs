#![deny(unused_must_use)]

mod ast;
mod event;
mod generated;
mod language;
mod lex;
mod marker;
mod parser;
mod precedence;
mod string_block;
mod tests;
mod token_set;

pub use ast::{AstChildren, AstNode, AstToken};
use event::Sink;
use generated::nodes::SourceFile;
pub use generated::{nodes, syntax_kinds::SyntaxKind};
pub use language::{
	JsonnetLanguage, PreorderWithTokens, SyntaxElement, SyntaxElementChildren, SyntaxNode,
	SyntaxNodeChildren, SyntaxToken,
};
use lex::lex;
use parser::{Parser, SyntaxError};
pub fn parse(input: &str) -> (SourceFile, Vec<SyntaxError>) {
	let lexemes = lex(input);
	let parser = Parser::new(&lexemes);
	let events = parser.parse();
	let sink = Sink::new(events, &lexemes);

	let parse = sink.finish();
	(
		SourceFile {
			syntax: parse.syntax(),
		},
		parse.errors,
	)
}
