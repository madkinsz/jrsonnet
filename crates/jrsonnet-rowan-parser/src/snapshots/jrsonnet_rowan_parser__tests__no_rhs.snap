---
source: crates/jrsonnet-rowan-parser/src/lib.rs
assertion_line: 88
expression: "a +\n"

---
Root@0..4
  BinOp@0..4
    Ident@0..2
      Ident@0..1 "a"
      Whitespace@1..2 " "
    OpPlus@2..3 "+"
    Whitespace@3..4 "\n"
===
Missing { expected: Named("value"), offset: 3 }
===
  [38;2;255;30;30mÃ—[0m syntax error
   â•­â”€â”€â”€â”€
 [2m1[0m â”‚ a +
   Â· [38;2;246;87;248m   â–²[0m
   Â·    [38;2;246;87;248mâ•°â”€â”€ [38;2;246;87;248mmissing value[0m[0m
   â•°â”€â”€â”€â”€

